File: mobile/lib/features/assets/data/models/asset_model.dart
import 'package:mobile/features/assets/domain/entities/asset.dart';

class AssetModel extends Asset {
  const AssetModel({
    required super.id,
    required super.assetNumber,
    required super.name,
    required super.status,
    super.companyId,
    super.category,
    super.serialNumber,
    super.manufacturer,
    super.model,
    super.installationDate,
    super.warrantyExpiry,
    super.location,
    super.gps,
    super.qrCode,
    super.barcode,
    super.customer,
    super.notes,
    super.image,
    super.createdAt,
    super.updatedAt,
  });

  factory AssetModel.fromJson(Map<String, dynamic> json) {
    final categoryJson = json['category'];
    final locationJson = json['location'];
    final gpsJson = json['gps'];
    final imageJson = json['image'];

    return AssetModel(
      id: json['id']?.toString() ?? '',
      companyId: json['companyId']?.toString(),
      assetNumber: json['assetNumber']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      category: categoryJson is Map<String, dynamic>
          ? AssetCategoryRef(
              id: categoryJson['id']?.toString() ?? '',
              name: categoryJson['name']?.toString(),
              code: categoryJson['code']?.toString(),
              icon: categoryJson['icon']?.toString(),
            )
          : null,
      serialNumber: json['serialNumber']?.toString(),
      manufacturer: json['manufacturer']?.toString(),
      model: json['model']?.toString(),
      installationDate: DateTime.tryParse(json['installationDate']?.toString() ?? ''),
      warrantyExpiry: DateTime.tryParse(json['warrantyExpiry']?.toString() ?? ''),
      status: AssetStatus.fromApi(json['status']?.toString()),
      location: locationJson is Map<String, dynamic>
          ? AssetLocation(
              branchId: locationJson['branchId']?.toString(),
              regionId: locationJson['regionId']?.toString(),
              cityId: locationJson['cityId']?.toString(),
              branchName: locationJson['branchName']?.toString(),
              regionName: locationJson['regionName']?.toString(),
              cityName: locationJson['cityName']?.toString(),
            )
          : const AssetLocation(),
      gps: gpsJson is Map<String, dynamic>
          ? AssetGps(
              latitude: (gpsJson['latitude'] as num?)?.toDouble(),
              longitude: (gpsJson['longitude'] as num?)?.toDouble(),
              accuracy: (gpsJson['accuracy'] as num?)?.toDouble(),
              address: gpsJson['address']?.toString(),
            )
          : const AssetGps(),
      qrCode: json['qrCode']?.toString(),
      barcode: json['barcode']?.toString(),
      customer: json['customer']?.toString(),
      notes: json['notes']?.toString(),
      image: imageJson is Map<String, dynamic>
          ? AssetImage(
              url: imageJson['url']?.toString() ?? '',
              publicId: imageJson['publicId']?.toString(),
              fileName: imageJson['fileName']?.toString(),
              mimeType: imageJson['mimeType']?.toString(),
              uploadedAt:
                  DateTime.tryParse(imageJson['uploadedAt']?.toString() ?? ''),
            )
          : null,
      createdAt: DateTime.tryParse(json['createdAt']?.toString() ?? ''),
      updatedAt: DateTime.tryParse(json['updatedAt']?.toString() ?? ''),
    );
  }
}

File: backend/src/modules/business/assets/models/asset.model.js
import mongoose from 'mongoose';

const { Schema } = mongoose;

export const ASSET_STATUSES = Object.freeze([
  'ACTIVE',
  'MAINTENANCE',
  'OFFLINE',
  'RETIRED',
]);

const imageSchema = new Schema(
  {
    url: { type: String, required: true },
    publicId: { type: String, default: null },
    fileName: { type: String, default: null },
    mimeType: { type: String, default: null },
    uploadedAt: { type: Date, default: Date.now },
  },
  { _id: false }
);

const gpsSchema = new Schema(
  {
    latitude: { type: Number, default: null },
    longitude: { type: Number, default: null },
    accuracy: { type: Number, default: null },
    address: { type: String, trim: true, default: null, maxlength: 500 },
  },
  { _id: false }
);

const locationSchema = new Schema(
  {
    branchId: { type: Schema.Types.ObjectId, ref: 'Branch', default: null },
    regionId: { type: Schema.Types.ObjectId, ref: 'Region', default: null },
    cityId: { type: Schema.Types.ObjectId, ref: 'City', default: null },
    branchName: { type: String, trim: true, default: null, maxlength: 200 },
    regionName: { type: String, trim: true, default: null, maxlength: 200 },
    cityName: { type: String, trim: true, default: null, maxlength: 200 },
  },
  { _id: false }
);

const assetSchema = new Schema(
  {
    companyId: {
      type: Schema.Types.ObjectId,
      ref: 'Company',
      required: true,
      index: true,
    },
    assetNumber: {
      type: String,
      required: true,
      trim: true,
      uppercase: true,
      maxlength: 100,
    },
    name: {
      type: String,
      required: true,
      trim: true,
      maxlength: 200,
    },
    categoryId: {
      type: Schema.Types.ObjectId,
      ref: 'AssetCategory',
      default: null,
      index: true,
    },
    serialNumber: {
      type: String,
      trim: true,
      default: null,
      maxlength: 120,
      index: true,
    },
    manufacturer: {
      type: String,
      trim: true,
      default: null,
      maxlength: 200,
    },
    model: {
      type: String,
      trim: true,
      default: null,
      maxlength: 200,
    },
    installationDate: {
      type: Date,
      default: null,
    },
    warrantyExpiry: {
      type: Date,
      default: null,
      index: true,
    },
    status: {
      type: String,
      enum: ASSET_STATUSES,
      default: 'ACTIVE',
      index: true,
    },
    location: {
      type: locationSchema,
      default: () => ({}),
    },
    gps: {
      type: gpsSchema,
      default: () => ({}),
    },
    qrCode: {
      type: String,
      trim: true,
      default: null,
      maxlength: 200,
      index: true,
    },
    barcode: {
      type: String,
      trim: true,
      default: null,
      maxlength: 200,
      index: true,
    },
    customer: {
      type: String,
      trim: true,
      default: null,
      maxlength: 200,
    },
    notes: {
      type: String,
      trim: true,
      default: null,
      maxlength: 5000,
    },
    image: {
      type: imageSchema,
      default: null,
    },
    deletedAt: {
      type: Date,
      default: null,
      index: true,
    },
  },
  {
    timestamps: true,
    collection: 'assets',
  }
);

assetSchema.index(
  { companyId: 1, assetNumber: 1 },
  { unique: true, partialFilterExpression: { deletedAt: null } }
);

assetSchema.index({ companyId: 1, name: 1, deletedAt: 1 });
assetSchema.index({ companyId: 1, status: 1, deletedAt: 1 });
assetSchema.index({ companyId: 1, 'location.branchId': 1, deletedAt: 1 });

const Asset = mongoose.model('Asset', assetSchema);

export default Asset;
